sdks/java/core/src/main/java/org/apache/beam/sdk/Pipeline.java

```diff
@@ -636,4 +649,49 @@ public void visitPrimitiveTransform(Node node) {
       node.getTransform().validate(options);
     }
   }
+
+  private static class TransformToMessage implements Function<PTransform<?, ?>, String> {
+    @Override
+    public String apply(final PTransform<?, ?> transform) {
+      final Object representant;
+      if (ParDo.SingleOutput.class.isInstance(transform)) {
```

Comment:
Hmm why are we special-casing ParDo here - why is line 664 ok for other transforms but not OK for ParDo.Single/MultiOutput? If it's a matter of the transform's .toString() - can you just change the toString() methods on these ParDo classes if they're currently not good enough?

Reply:
To get a user data and not an internal beam data. ParDo doesnt help here for instance.

Reply:
Sorry, I don't understand which internal Beam data you have in mind. Line 664 calls toString() - so all you need to get rid of special-casing ParDo as opposed to the dozens of other Beam transforms is to add .toString() methods to ParDo.SingleOutput and MultiOutput, which will be useful for debugging anyway.

Reply:
Hmm, not really cause the goal is to log the min info needed for the user to fix it. XOutput doesnt help since it is ignored in the name computation.

Reply:
If I understand correctly: the default .toString() on a ParDo returns simply "ParDo.SingleOutput@something" or alike; and instead you want to return a .toString() of the DoFn inside it. Can you just override the toString() method on the ParDo.SingleOutput class to say "return fn.toString()"?

Reply:
Sounds a bit weird to bypass a wrapper but Im fine with it. Will push it tmr. Thanks for the comment/details.